src/test/java/com/make/my/day/hm5/MergeSortJoinTest.java

```diff
@@ -104,31 +99,137 @@ public void spliteratorMemoryTest() {
 
     //ToDo: Implement your own merge sort inner join spliterator. See https://en.wikipedia.org/wiki/Sort-merge_join
     public static class MergeSortInnerJoinSpliterator<C extends Comparable<C>, L, R> implements Spliterator<Pair<L, R>> {
+        Stream<L> leftStream;
+        Stream<R> rightStream;
+        Function<L, C> keyExtractorLeft;
+        Function<R, C> keyExtractorRight;
+        L nextLeft;
+        R nextRight;
+        Iterator<L> ltIter;
+        Iterator<R> rIter;
+        ArrayList<L> lBuffer = new ArrayList<>();
+        ArrayList<R> rBuffer = new ArrayList<>();
+        Short flag;
+        Short leftInd;
+        Short rightInd;
+
         public MergeSortInnerJoinSpliterator(Stream<L> left,
                                              Stream<R> right,
                                              Function<L, C> keyExtractorLeft,
                                              Function<R, C> keyExtractorRight,
                                              boolean isSorted) {
+            if (!isSorted) {
+                this.ltIter = left.sorted().iterator();
+                this.rIter = right.sorted().iterator();
+            } else {
+                this.ltIter = left.iterator();
+                this.rIter = right.iterator();
+            }
+            this.leftStream = left;
+            this.rightStream = right;
+            this.keyExtractorLeft = keyExtractorLeft;
+            this.keyExtractorRight = keyExtractorRight;
+            this.nextLeft = ltIter.next();
```

Comment:
What if ltIter doesn't have next element?

Reply:
Thank you for question, I added a check for this border point!